### PR TITLE
Madison/rationalized

### PR DESCRIPTION
**Pull request by madisonmay**

### CODE REVIEW

File: enso/experiment/rationalized.py [hunk 1]
```diff
@@ -92,4 +95,122 @@ def predict(self, X, **kwargs):
...
+    def _valid(self, token):
+        return (
+            token.has_vector 
+            and not token.is_stop 
+            and np.any(np.nonzero(token.vector))
+        )
+
+    def _featurize(self, doc):
+        doc_vect = np.mean([
+            token.vector * self._rationale_weight(token)
```
Comment (madisonmay) [inline enso/experiment/rationalized.py]: Yes I should probably vectorize this

Comment (madisonmay) [reply enso/experiment/rationalized.py]: Awkwardly I did and it was... slower.

```
def _featurize(self, doc):
        doc_vectors = np.vstack([token.vector for token in doc if self._valid(token)])
        dot_products = np.dot(doc_vectors, self.normalized_rationale_proto) 
        rationale_weights = (1. + dot_products / np.linalg.norm(doc_vectors, axis=1)) / 2.
        doc_vector = np.sum(rationale_weights.reshape(-1, 1) * doc_vectors, axis=0)
        return doc_vector
```

File: enso/experiment/rationalized.py [hunk 2]
```diff
@@ -92,4 +95,211 @@ def predict(self, X, **kwargs):
...
+            self.p_rationale_given_word[word] = np.sqrt(count / word_counts.get(word, count))
+
+        return self.p_rationale_given_word
+
+    def _featurize(self, doc):
+        doc_vect = np.mean([
+            token.vector * self.p_rationale_given_word[token.text.lower()]
+            for token in doc
+        ], axis=0)
+        normed_doc_vect = doc_vect / np.linalg.norm(doc_vect)
```
Comment (benleetownsend) [inline enso/experiment/rationalized.py]: Have you tried it without this normalisation?

**Comment by benleetownsend** [comment]
All this checks out, 2 questions I have are 
1) do the norms seem to help/hurt performance.
2) do you have some metrics on the rationale predictions themselves? Is this bit actually a tractable task with the current approaches?

File: enso/experiment/rationalized.py (hunk 2 above)
Comment (madisonmay) [reply enso/experiment/rationalized.py]: Nope but I should for sure

**Pull request merged**